pex/orderedset.py

```diff
@@ -10,71 +10,65 @@
 
 from __future__ import absolute_import
 
+from collections import OrderedDict
+
 from pex.compatibility import MutableSet
+from pex.typing import TYPE_CHECKING, Generic
+
+if TYPE_CHECKING:
+    from typing import Any, Iterable, Optional, TypeVar
 
+    _I = TypeVar("_I")
 
-class OrderedSet(MutableSet):
-    KEY, PREV, NEXT = range(3)
 
+class OrderedSet(MutableSet, Generic["_I"]):
     def __init__(self, iterable=None):
-        self.end = end = []
-        end += [None, end, end]  # sentinel node for doubly linked list
-        self.map = {}  # key --> [key, prev, next]
+        # type: (Optional[Iterable[_I]]) -> None
+        self.map = OrderedDict()  # type: OrderedDict[_I, None]
```

Comment:
If the API is changing anyway (public `self.end` going away), it seems like maybe a good time to hide `self.map`, unless it's consumed within the codebase?

Reply:
SGTM - I changed it back from self._data to self.map at the last minute just to be nice to reviewers.